core/src/main/java/org/apache/iceberg/io/BaseTaskWriter.java

```diff
@@ -75,6 +81,135 @@ public WriteResult complete() throws IOException {
         .build();
   }
 
+  /**
+   * Base equality delta writer to write both insert records and equality-deletes.
+   */
+  protected abstract class BaseEqualityDeltaWriter implements Closeable {
+    private final StructProjection structProjection;
+    private RollingFileWriter dataWriter;
+    private RollingEqDeleteWriter eqDeleteWriter;
+    private SortedPosDeleteWriter<T> posDeleteWriter;
+    private Map<StructLike, PathOffset> insertedRowMap;
+
+    public BaseEqualityDeltaWriter(PartitionKey partition, Schema schema, Schema deleteSchema) {
+      Preconditions.checkNotNull(schema, "Iceberg table schema cannot be null.");
+      Preconditions.checkNotNull(deleteSchema, "Equality-delete schema cannot be null.");
+      this.structProjection = StructProjection.create(schema, deleteSchema);
+
+      this.dataWriter = new RollingFileWriter(partition);
+
+      this.eqDeleteWriter = new RollingEqDeleteWriter(partition);
+      this.insertedRowMap = StructLikeMap.create(deleteSchema.asStruct());
+
+      this.posDeleteWriter = new SortedPosDeleteWriter<>(appenderFactory, fileFactory, format, partition);
+    }
+
+    /**
+     * Make the generic data could be read as a {@link StructLike}.
+     */
+    protected abstract StructLike asStructLike(T data);
+
+    public void write(T row) throws IOException {
+      PathOffset pathOffset = PathOffset.of(dataWriter.currentPath(), dataWriter.currentRows());
+
+      StructLike copiedKey = structProjection.copy().wrap(asStructLike(row));
+      // Adding a pos-delete to replace the old path-offset.
+      PathOffset previous = insertedRowMap.put(copiedKey, pathOffset);
```

Comment:
Doesn't this have the problem you were talking about with using only wrappers? If `asStructLike` uses a wrapper that is reused, then even if the key is a new instance, the underlying data in the `StructLike` wrapper returned will change on the next call to write.

I think instead of copying the projection, this needs to copy the result of the projection into a new struct and add that to the map.

Reply:
I'm a little hesitant about the `asCopiedStructLike`,  because it introduces complexity to the compute engine developer, people need to consider in which case they need to use `asStructLike` other cases they need to use `asCopiedStructLike`.  Implementing the two abstracted method will need carefully coding.   So I'm thinking how about always copy the `StructLike`  ?   see the flink implementation [here](https://github.com/apache/iceberg/pull/1896/files#diff-e497cabcc9fcf3b1edcd987693853c2dd578f0b3231d5a8f4c3b6f0338bf5e3cR95). 

The copy is light-weight because only few references copy ( from `RowDataWrapper` ).  That would simplify the abstracted methods design.

Reply:
Finally, I decided to introduce the `asCopiedStructLike`.  People don't have to care how to use it inside the `BaseEqualityDeltaWriter`,  just need to know we need to implement it as a totally new StructLike which won't effect the old one.

Reply:
Why did you decide to delegate to the implementation instead of creating a copy here? I think it would be fairly easy to copy with a StructLike class:

```java
  public static class StructCopy implements StructLike {
    public static StructLike copy(StructLike struct) {
      return new StructCopy(struct);
    }

    private final Object[] values;

    private StructCopy(StructLike toCopy) {
      this.values = new Object[toCopy.size()];
      for (int i = 0; i < values.length; i += 1) {
        Object value = toCopy.get(i, Object.class);
        if (value instanceof StructLike) {
          values[i] = copy((StructLike) value);
        } else {
          values[i] = value;
        }
      }
    }

    @Override
    public int size() {
      return values.length;
    }

    @Override
    public <T> T get(int pos, Class<T> javaClass) {
      return javaClass.cast(values[pos]);
    }

    @Override
    public <T> void set(int pos, T value) {
      throw new UnsupportedOperationException("Struct copy cannot be modified");
    }
  }
```

That doesn't handle lists or maps, but the projection doesn't handle lists or maps either, so it would be okay to use that in the StructLikeMap for all of the cases we currently support.

Reply:
Okay, I misunderstood your point. Sounds good to create a `StructCopy` to copy the values, I like the idea. Thanks.